policies/coding-style.md

```diff
@@ -584,28 +594,99 @@ less complicated than coding pure assembly, and it doesn't provide the
 same performance guarantee across different micro-architecture. Nor is
 it portable enough to meet our multi-platform support goals.
 
-## Chapter 14:  Portability
+## Chapter 14: Portability
 
 To maximise portability the version of C defined in ISO/IEC 9899:1990
 should be used. This is more commonly referred to as C90. ISO/IEC 9899:1999
 (also known as C99) is not supported on some platforms that OpenSSL is
 used on and therefore should be avoided.
 
-## Chapter 15: Miscellaneous
+## Chapter 15: Expressions
 
-Do not use `!` to check if a pointer is NULL, or to see if a str...cmp
-function found a match.  For example, these are wrong:
+Avoid needless parentheses as far as reasonable.
+For example, do not write
+```c
+    if ((p == NULL) && (!f(((2 * x) + y) == (z++))))
+```
+but
+```c
+    if (p == NULL && !f(2 * x + y == z++)).
+```
 
+For clarity, always put parenthese
+when mixing the logical `&&` and `||` operators,
+mixing comparison operators like `<=` and `==,`
+or mixing bitwise operators like  `&` and `|` .
+For example,
 ```c
-    if (!(p = BN_new())) ...
-    if (!strcmp(a, "FOO")) ...
+    if ((a && b) || c)
+    if ((a <= b) == ((c >= d) != (e < f)))
+    x = (a & b) ^ (c | d)
 ```
 
-Do this instead:
+Regarding parentheses in macro definitions see the
+[chapter on macros](#chapter-9-macros-and-enums).
 
+In comparisons with constants (including `NULL` and other constant macros)
+place the constant on the right-hand side of the comparison operator.
+For example,
 ```c
-    if ((p = BN_new()) == NULL)...
-    if (strcmp(a, "FOO") == 0) ...
+    while (i++ < 10 && p != NULL)
+```
+
+Do not use implicit checks for
+numbers (not) being `0` or pointers (not) being `NULL`.
+For example, do not write
+```c
+    if (i)
+    if (!(x & MASK))
+    if (!strcmp(a, "FOO"))
+    if (!(p = BN_new()))
+```
+but do this instead:
+```c
+    if (i != 0)
+    if ((x & MASK) == 0)
+    if (strcmp(a, "FOO") == 0)
+    if ((p = BN_new()) == NULL)
+```
+Boolean values shall be used directly as usual, e.g.,
+```c
+if (check(x) && !success(y))
+```
+
```

Comment:
What about a note: _Many functions can return 0 or a negative value on error and the Boolean forms need to be used with care._? This is a very common error in the code base.

Reply:
Yeah, good to add such a caveat - done.